### PR TITLE
Client: write files to disk and verify checksums

### PR DESCRIPTION
**Pull request by hendrikcech**
Added -o / --out flag that can also be set to '-' to write the received data to stdout.

In addition, the user of rft (cmd.go) must use ChecksumValid() once a file is fully transmitted.

### CODE REVIEW

**Comment by hendrikcech** [comment]
Related issue is #22.

File: cmd/cmd.go [hunk 1]
```diff
@@ -70,14 +76,32 @@ var rootCmd = &cobra.Command{
...
 		if err != nil {
 			log.Printf("error on request: %v\n", err)
 		}
 
-		for _, r := range rs {
-			io.Copy(os.Stdout, r)
-			log.Println("finish write")
+		for i, req := range reqs {
+			var w io.Writer
+			// TODO: verify checksum: method on req?
```
Comment (mengelbart) [inline cmd/cmd.go:86]: TODO seems done? Alternatively, call checksum after rft is done with all bytes and if it doesn't match, add an entry to the error field of the response?

File: rftp/file_response.go [hunk 2]
```diff
@@ -45,11 +49,27 @@ func newFileResponse(index uint16) *FileResponse {
...
 		resendEntries: make(map[uint64]struct{}),
+		hasher:        md5.New(),
 	}
 }
 
 func (f *FileResponse) Read(p []byte) (n int, err error) {
-	return f.preader.Read(p)
+	n, err = f.preader.Read(p)
+	if err != nil {
+		return
```
Comment (mengelbart) [inline rftp/file_response.go:59]: This `return` comes too early. `err` might be EOF and the documentation of [io.Reader](https://golang.org/pkg/io/#Reader) says you have to handle n>0 bytes before considering the error. Thus, we have to add n bytes to the hasher before returning the error, which may be EOF or some other io error.

File: rftp/file_response.go [hunk 3]
```diff
@@ -45,11 +49,27 @@ func newFileResponse(index uint16) *FileResponse {
...
+	}
+	_, err = f.hasher.Write(p[:n])
+	if err != nil {
```
Comment (mengelbart) [inline rftp/file_response.go:62]: I think this if is unnecessary, in both cases you return the same `n, err`.

File: rftp/file_response.go [hunk 4]
```diff
@@ -45,11 +49,27 @@ func newFileResponse(index uint16) *FileResponse {
...
+		return
+	}
+	return
+}
+
+// Returns true iff the checksum of the received file matches the checksum
+// transmitted in the metadata message. Returns false while the file is still in
+// transit.
+func (f *FileResponse) ChecksumValid() bool {
+	return bytes.Compare(f.checksum[:], f.hasher.Sum(nil)[:16]) == 0
```
Comment (mengelbart) [inline rftp/file_response.go:72]: The [`bytes.Compare`](https://golang.org/pkg/bytes/#Compare) documentation says *Prefer Equal to Compare for equality comparisons.* Is there a reason why you use `Compare` anyway?

File: cmd/cmd.go [hunk 5]
```diff
@@ -132,13 +156,23 @@ func directoryHandler(dirname string) (rftp.FileHandler, error) {
...
+
+	rootCmd.PersistentFlags().Float32VarP(&p, "p", "p", -1, `specify the loss
+probabilities for the Markov chain model (0 <= p <= 1) if only one is specified,
+assume p=q; if neither is specified assume no loss`)
+
+	rootCmd.PersistentFlags().Float32VarP(&q, "q", "q", -1, `specify the loss
+probabilities for the Markov chain model (0 <= p <= 1) if only one is specified,
+assume p=q; if neither is specified assume no loss`)
+
+	rootCmd.PersistentFlags().StringVarP(&out, "out", "o", ".", `specify the
```
Comment (mengelbart) [inline cmd/cmd.go:173]: I think I understand the line breaks for most of the flags and I think they make sense, but the breaks in the `out` flag look wrong to me:
2020-07-22-194415_grim

File: rftp/file_response.go [hunk 6]
```diff
@@ -109,16 +134,19 @@ func (f *FileResponse) write(done chan<- uint16) {
 	log.Printf("Start processing file %v\n", f.index)
 	defer func() {
 		done <- f.index
-		f.pwriter.Close()
+		// If simply Close() (or equivalently CloseWithError(nil) is called, preader
+		// never returns err=io.EOF and cmd.go hangs in io.Copy(w, req). Test case:
+		// request not-existing file.
+		f.pwriter.CloseWithError(dummyError)
+		// f.pwriter.Close()
```
Comment (hendrikcech) [inline rftp/file_response.go:141]: That is really weird. Do you know why it behaves like that? The solution is definitely not optimal..

Comment (mengelbart) [reply rftp/file_response.go:141]: Actually I can't reproduce this. If I switch to just `f.pwriter.Close()`, I get this log:
```
2020/07/23 11:10:52 file 0 at head 11 and buffer size 0
2020/07/23 11:10:52 Finished processing file 0
2020/07/23 11:10:52 fr Read: err EOF
2020/07/23 11:10:52 File testfile received (checksum is valid)
2020/07/23 11:10:52 send abort to file writer: 0
2020/07/23 11:10:52 leaving ack writer
```
This is fine, isn't it?

Comment (hendrikcech) [reply rftp/file_response.go:141]: Try requesting a file that does not exist. In case the file exists, I also get the expected behavior.

Comment (mengelbart) [reply rftp/file_response.go:141]: I see. But I get the same behavior for both, `Close` and `CloseWithError`. But I did some tests and I think it's a deadlock in the `Read` method, where you try to acquire a lock to set `f.Err`. But that lock has previously been acquired and not been freed by `Write`, in the metadata case of the select, when it returns early, after encountering the error status.

Comment (hendrikcech) [reply rftp/file_response.go]: Oh wow, good catch!

**Pull request merged**